modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/BlobStorage.java

```diff
@@ -0,0 +1,341 @@
+ * (probably using a {@link org.apache.ignite.internal.pagememory.freelist.FreeList}) should be used.
+ */
+public class BlobStorage {
+    static final long NO_PAGE_ID = 0;
+
+    private final ReuseList reuseList;
+    private final PageMemory pageMemory;
+
+    private final int groupId;
+    private final int partitionId;
+
+    private final IoStatisticsHolder statisticsHolder;
+
+    private final RecycleAndAddToReuseBag recycleAndAddToReuseBag = new RecycleAndAddToReuseBag();
+
+    private final ReadFragment readFragment = new ReadFragment();
+
+    private final WriteFragment writeFragment = new WriteFragment();
+
+    /**
+     * Creates a new instance.
+     */
+    public BlobStorage(ReuseList reuseList, PageMemory pageMemory, int groupId, int partitionId, IoStatisticsHolder statisticsHolder) {
+        this.reuseList = reuseList;
+        this.pageMemory = pageMemory;
+        this.groupId = groupId;
+        this.partitionId = partitionId;
+        this.statisticsHolder = statisticsHolder;
+    }
+
+    /**
+     * Reads a blob stored starting at a page with the given ID.
+     *
+     * @param firstPageId ID of first page.
+     * @return Byte array for the blob.
+     * @throws IgniteInternalCheckedException If something goes wrong.
+     */
+    public byte[] readBlob(long firstPageId) throws IgniteInternalCheckedException {
+        ReadState readState = new ReadState();
+
+        long pageId = firstPageId;
+
+        while (pageId != NO_PAGE_ID) {
+            Boolean ok = PageHandler.readPage(
+                    pageMemory,
+                    groupId,
+                    pageId,
+                    PageLockListenerNoOp.INSTANCE,
+                    readFragment,
+                    readState,
+                    0,
+                    false,
+                    IoStatisticsHolderNoOp.INSTANCE
+            );
+
+            assert ok : pageId;
+
+            pageId = readState.nextPageId;
+        }
+
+        assert readState.bytes != null;
+
+        return readState.bytes;
+    }
+
+    /**
+     * Adds a new blob to the storage.
+     *
+     * @param bytes Blob bytes.
+     * @return ID of the page starting the chain representing the blob.
+     * @throws IgniteInternalCheckedException If something goes wrong.
+     */
+    public long addBlob(byte[] bytes) throws IgniteInternalCheckedException {
+        return doStore(NO_PAGE_ID, bytes);
+    }
+
+    /**
+     * Updates the blob content.
+     *
+     * @param firstPageId ID of the first page in the chain storing the blob.
+     * @param bytes New blob content.
+     * @throws IgniteInternalCheckedException If something goes wrong.
+     */
+    public void updateBlob(long firstPageId, byte[] bytes) throws IgniteInternalCheckedException {
+        doStore(firstPageId, bytes);
+    }
+
+    private long doStore(long maybeFirstPageId, byte[] bytes) throws IgniteInternalCheckedException {
```

Comment:
What about the fact that we always free old pages and create new ones at the beginning? seems like a simpler option.

Reply:
This looks like a suggestion and not a fact :)

I wanted to avoid unneeded reallocation.

Reply:
How often will we update the configuration?

Reply:
I agree that we should reuse old pages, it's simpler actually. And I believe that for the most part, the amount of pages will stay the same.

Reply:
We'll update RAFT config on each configuration change, this should happen rarely. But it won't hurt to make the mechanism a bit more generic (it doesn't cost too much in terms of complexity).

Reply:
Ok